### PR TITLE
Added erf(x) Float64 Julia implementation

### PR DESCRIPTION
**Pull request by AhmedsElysium**
Faster than current wrapper function call (including Float32 function call).
Uses algorithm based on https://github.com/ARM-software/optimized-routines/blob/master/math/erf.c

### CODE REVIEW

**Comment by AhmedsElysium** [comment]
**Old:**
Float 64
`@benchmark SpecialFunctions.erf(data) setup=(data=6*rand(Float64)-3) samples=1000000`

> BenchmarkTools.Trial: 217729 samples with 1000 evaluations per sample.
>  Range (min … max):   6.300 ns … 283.700 ns  ┊ GC (min … max): 0.00% … 0.00%
>  Time  (median):     29.500 ns               ┊ GC (median):    0.00%
>  Time  (mean ± σ):   21.993 ns ±  13.209 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

Float32
`@benchmark SpecialFunctions.erf(data) setup=(data=6*rand(Float32)-3) samples=1000000`

> BenchmarkTools.Trial: 312732 samples with 1000 evaluations per sample.
>  Range (min … max):   4.300 ns … 125.100 ns  ┊ GC (min … max): 0.00% … 0.00%
>  Time  (median):     19.900 ns               ┊ GC (median):    0.00%
>  Time  (mean ± σ):   15.035 ns ±   7.951 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

**New:**
Float64
`@benchmark erf(data) setup=(data=6*rand(Float64)-3) samples=1000000`

> BenchmarkTools.Trial: 507504 samples with 1000 evaluations per sample.
>  Range (min … max):  5.400 ns … 4.890 μs  ┊ GC (min … max): 0.00% … 0.00%
>  Time  (median):     8.700 ns             ┊ GC (median):    0.00%
>  Time  (mean ± σ):   8.775 ns ± 9.855 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

Float32
`@benchmark Float32(erf(data)) setup=(data=6*rand(Float64)-3) samples=1000000`

> BenchmarkTools.Trial: 526797 samples with 1000 evaluations per sample.
>  Range (min … max):  5.400 ns … 195.500 ns  ┊ GC (min … max): 0.00% … 0.00%
>  Time  (median):     8.800 ns               ┊ GC (median):    0.00%
>  Time  (mean ± σ):   8.521 ns ±   2.236 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

**Comment by AhmedsElysium** [comment]
Float32 implementation available, but not faster than Float64 version due to a exp() call.
Float64 version still faster than old Float32.

**Comment by AhmedYKadah** [comment]
need to clean up polynomial evaluations. 
code also could use more organization

**Comment by AhmedYKadah** [comment]
Remaining: erfc Float64 and Float32 implementations, and the erf Float32 implementation

File: src/erf.jl [hunk 1]
```diff
@@ -96,10 +97,211 @@ See also:
...
+
+function _erf(x::Float32)
+    xabs=abs(x)
+
+  if (xabs< 0.5)
+    # range [0;0.5]
+    # # erf approximation using erf(x)=x+x*P(x^2) with degree 6
+    # Sollya::remez(erf(sqrt(x))/sqrt(x)-1,6,[1e-32;0.5],1,1e-32);
+
+    p=(0.128379167084072442015971708878153077794812875442854468252262977759538466466026108f0 ,-0.37612638677173360887708232154133809982218085424044576090554860522057182310851056f0 ,0.112837843774249243616280397250157777250507394914770984188915122197580134574091983f0 ,-2.68652865375831097248410803484628824158061934615600565461807890879374835870381413f-2 ,5.2188560068141289500478726517878958032487316358404909126782459380536157982980317f-3 ,-8.3948481565340715162646401327604259604058983841328238470236141987304564416187075f-4 )
```
Comment (oscardssmith) [inline src/erf.jl]: these are very long Float32 literals. If you copy and paste them into the repl, Julia will give you minimal length versions.

File: src/erf.jl [hunk 2]
```diff
@@ -96,10 +97,211 @@ See also:
...
+        else
+            return 1.0
+        end
+
+    end
+
+
+end
+
+function _erf(x::Float32)
```
Comment (oscardssmith) [inline src/erf.jl:214]: add comment for total accuracy like for the Float64 version

File: src/erf.jl [hunk 3]
```diff
@@ -96,10 +97,211 @@ See also:
...
+
+        r=evalpoly(a,PF)
+
+
+        if (sign)
+            return -1.0 + r
+        else
+            return 1.0 - r
+        end
+    else
```
Comment (oscardssmith) [inline src/erf.jl]: why not do this with `copysign` like in the Float32 version?

File: src/erf.jl [hunk 4]
```diff
@@ -96,10 +97,211 @@ See also:
...
+    ix::UInt32=reinterpret(UInt64,x)>>32
+    # # top 32, without sign bit 
+    ia::UInt32=ix & 0x7fffffff
+    # # sign
+    # sign::UInt32=ix>>31
+
+    sign::Bool=x<0
+
+
+
```
Comment (oscardssmith) [inline src/erf.jl]: remove some of the excess whitespace.

File: src/erf.jl [hunk 5]
```diff
@@ -96,10 +97,211 @@ See also:
...
+        r=evalpoly(a,PE)
+
+
+        if (sign)
+            return -1.0 + r
+        else
+            return 1.0 - r
+        end
+    elseif (ia < 0x4017a000)
```
Comment (oscardssmith) [inline src/erf.jl]: any chance you can regenerate the polies to make this be `0x40180000`? If you can, you would be able to use UInt16 literals for all of these by only taking the top 16 rather than top 32 bits.

Comment (AhmedYKadah) [reply src/erf.jl]: would that have a meaningful impact?

Comment (oscardssmith) [reply src/erf.jl]: probably not much. It might just save a cycle or 2.

File: src/erf.jl [hunk 6]
```diff
@@ -96,10 +97,165 @@ See also:
...
+        # Generated using Sollya::remez(f(c*x+d), deg, [(a-d)/c;(b-d)/c], 1, 1e-16), [|D ...|] with deg=17 a=2 b=3.25 c=2 d=2 
+        PD=( 0x1.328f5ec350e5p-8, -0x1.529b9e8cf8e99p-5, 0x1.529b9e8cd9e71p-3, -0x1.8b0ae3a023bf2p-2, 0x1.1a2c592599d82p-1, -0x1.ace732477e494p-2, -0x1.e1a06a27920ffp-6, 0x1.bae92a6d27af6p-2, -0x1.a15470fcf5ce7p-2, 0x1.bafe45d18e213p-6, 0x1.0d950680d199ap-2, -0x1.8c9481e8f22e3p-3, -0x1.158450ed5c899p-4, 0x1.c01f2973b44p-3, -0x1.73ed2827546a7p-3, 0x1.47733687d1ff7p-4, -0x1.2dec70d00b8e1p-6, 0x1.a947ab83cd4fp-10 )
+
+        # Obtains erfc of |x|
+        r=1.0-evalpoly(a,PD)
+        return copysign(r,x)
+
+    elseif (ia < 0x40100000)
+    ## 3.25 <= |x| < 4.0.  
+        a = abs(x)
```
Comment (oscardssmith) [inline src/erf.jl:185]: move the `abs(x)` out of the branches?

Comment (AhmedYKadah) [reply src/erf.jl:185]: It's not used in all branches

Comment (AhmedYKadah) [reply src/erf.jl:185]: wouldn't it add some latency?

File: src/erf.jl (hunk 2 above)
Comment (AhmedYKadah) [reply src/erf.jl:214]: done

File: src/erf.jl (hunk 3 above)
Comment (AhmedYKadah) [reply src/erf.jl]: done

File: src/erf.jl (hunk 4 above)
Comment (AhmedYKadah) [reply src/erf.jl]: done

File: src/erf.jl (hunk 5 above)
Comment (oscardssmith) [reply src/erf.jl]: if you wanted to test the speed, you could try it without regenerating the polys and it should give you a good idea.

File: src/erf.jl [hunk 7]
```diff
@@ -256,47 +219,38 @@ function _erf(x::Float32)
...
-  return r
-
+    # range [4.0,inf)
+    r=copysign(1f0,x)
+    return r
   end
-
 end
 
 _erf(x::Float16)=Float16(_erf(Float32(x)))
```
Comment (oscardssmith) [inline src/erf.jl:257]: if you wanted to do a `Float16` impl, it should be easier than the others. Specifically, the domain is only to 2, and the accuracy required is much reduced.

Comment (oscardssmith) [reply src/erf.jl:257]: 100% could wait for a followup PR.

File: src/erf.jl (hunk 1 above)
Comment (AhmedYKadah) [reply src/erf.jl]: done

File: src/erf.jl (hunk 7 above)
Comment (AhmedYKadah) [reply src/erf.jl:257]: I'm thinking that too to be honest.
this and the poli regen.

**Comment by mschauer** [comment]
Are there any tests for edge cases/ULP in the c version we do not do ourselves?

File: src/erf.jl [hunk 8]
```diff
@@ -96,10 +97,165 @@ See also:
...
+        PF=( 0x1.08ddd130d1fa6p-26, -0x1.10b146f59ff06p-22, 0x1.10b135328b7b2p-19, -0x1.6039988e7575fp-17, 0x1.497d365e19367p-15, -0x1.da48d9afac83ep-14, 0x1.1024c9b1fbb48p-12, -0x1.fc962e7066272p-12, 0x1.87297282d4651p-11, -0x1.f057b255f8c59p-11, 0x1.0228d0eee063p-10, -0x1.b1b21b84ec41cp-11, 0x1.1ead8ae9e1253p-11, -0x1.1e708fba37fccp-12, 0x1.9559363991edap-14, -0x1.68c827b783d9cp-16, 0x1.2ec4adeccf4a2p-19 )
+
+        r=1.0-evalpoly(a,PF)
+        return copysign(r,x)
+    else
+        if(isnan(x))
+             return NaN
+        end
+        copysign(1.0,x)
+    end
```
Comment (devmotion) [inline src/erf.jl:208]: Maybe add an explicit return statement to the fallback:
```suggestion
    elseif isnan(x)
        return NaN
    else
        return copysign(1.0,x)
    end
```

File: src/erf.jl [hunk 9]
```diff
@@ -96,10 +97,165 @@ See also:
...
+    # Sollya::remez(erfc(x+2.5),13,[0;4-2.5],1,1e-32);
+
+    p=(0.00040695202f0, -0.002178284f0, 0.0054457085f0, -0.008350053f0, 0.008622011f0, -0.006115167f0, 0.0027899458f0, -0.000519395f0, -0.00030461047f0, 0.00031068458f0, -0.00013866898f0, 3.6909692f-5, -5.682889f-6, 3.929763f-7)
+    return copysign(1f0-evalpoly(xabs-2.5f0,p),x)
+    
+  else
+    # range [4.0,inf)
+    r=copysign(1f0,x)
+    return r
+  end
```
Comment (devmotion) [inline src/erf.jl:254]: This misses the `NaN` case:
```suggestion
  elseif isnan(x)
    return NaN
  else
    # range [4.0,inf)
    return copysign(1f0, x)
  end
```

File: src/erf.jl [hunk 10]
```diff
@@ -28,6 +25,10 @@ for f in (:erf, :erfc)
     end
 end
 
+    _erfc(x::Float64) = ccall((:erfc, libopenlibm), Float64, (Float64,), x)
+    _erfc(x::Float32) = ccall((:erfcf, libopenlibm), Float32, (Float32,), x)
+    _erfc(x::Float16) = Float16(_erfc(Float32(x)))
```
Comment (devmotion) [inline src/erf.jl]:
```suggestion
_erfc(x::Float64) = ccall((:erfc, libopenlibm), Float64, (Float64,), x)
_erfc(x::Float32) = ccall((:erfcf, libopenlibm), Float32, (Float32,), x)
_erfc(x::Float16) = Float16(_erfc(Float32(x)))
```

**Comment by mschauer** [comment]
Then we should also add a test for these

File: src/erf.jl [hunk 11]
```diff
@@ -96,10 +97,165 @@ See also:
...
+  elseif (xabs<4.0)
+    # range [2.5;4.0]
+    # erfc approximation with degree 13
+    # Sollya::remez(erfc(x+2.5),13,[0;4-2.5],1,1e-32);
+
+    p=(0.00040695202f0, -0.002178284f0, 0.0054457085f0, -0.008350053f0, 0.008622011f0, -0.006115167f0, 0.0027899458f0, -0.000519395f0, -0.00030461047f0, 0.00031068458f0, -0.00013866898f0, 3.6909692f-5, -5.682889f-6, 3.929763f-7)
+    return copysign(1f0-evalpoly(xabs-2.5f0,p),x)
+    
+  elseif isnan(x)
+    return NaN
```
Comment (devmotion) [inline src/erf.jl]:
```suggestion
    return NaN32
```

**Comment by AhmedYKadah** [comment]
Any other changes needed?

**Comment by oscardssmith** [comment]
we should probably should test erfc.

File: test/erf.jl [hunk 12]
```diff
@@ -2,10 +2,46 @@
...
+            @test erfc(T(-0.25))   ≈ T(1.276326390168237) rtol=2*eps(T)
+            @test erfc(T(-0.75))   ≈ T(1.7111556336535152) rtol=2*eps(T)
+            @test erfc(T(-1.0))   ≈ T(1.8427007929497148) rtol=2*eps(T)
+            @test erfc(T(-1.5))   ≈ T(1.9661051464753108) rtol=2*eps(T)
+            @test erfc(T(-2.5))   ≈ T(1.999593047982555) rtol=2*eps(T)
+            @test erfc(T(-3.5))   ≈ T(1.9999992569016276) rtol=2*eps(T)
+            @test erfc(T(-4.5))   ≈ T(1.999999999803384) rtol=2*eps(T)
+            @test erfc(T(-6.0))   ≈ T(2) rtol=2*eps(T)
+
+            @test isnan(erfc(T(NaN)))
```
Comment (oscardssmith) [inline test/erf.jl:45]: test Inf?

**Comment by oscardssmith** [comment]
Other than missing tests for Inf, looks good to me. @devmotion any further sugestions?

File: src/erf.jl [hunk 13]
```diff
@@ -96,10 +97,165 @@ See also:
 [`erfinv(x)`](@ref erfinv), [`erfcinv(x)`](@ref erfcinv).
 
 # Implementation by
-- `Float32`/`Float64`: C standard math library
-    [libm](https://en.wikipedia.org/wiki/C_mathematical_functions#libm).
+- `Float32`: polynomial approximations of erf 
+- `Float64`: Julia implementation of https://github.com/ARM-software/optimized-routines/blob/master/math/erf.c 
 - `BigFloat`: C library for multiple-precision floating-point [MPFR](https://www.mpfr.org/)
 """
+
```
Comment (devmotion) [inline src/erf.jl:105]: This breaks (ie removes) the docstring of `erf`.

Comment (devmotion) [reply src/erf.jl:105]:
```suggestion
function erf end

```

File: src/erf.jl [hunk 14]
```diff
@@ -96,10 +97,165 @@ See also:
...
+  else
+    # range [4.0,inf)
+    return copysign(1f0, x)
+  end
+end
+
+_erf(x::Float16)=Float16(_erf(Float32(x)))
+
+
 function erf end
```
Comment (devmotion) [inline src/erf.jl]:
```suggestion
```

File: src/erf.jl (hunk 5 above)
Comment (AhmedYKadah) [reply src/erf.jl]: Done

File: src/erf.jl (hunk 13 above)
Comment (AhmedYKadah) [reply src/erf.jl:105]: Could you check it after the new commit?

**Comment by AhmedYKadah** [comment]
Yes, you're correct. The old implementation was not efficient so I redid it.

# Old
**Float64**
`@benchmark erf(data) setup=(data=6*rand(Float64)-3) samples=1000000`

> BenchmarkTools.Trial: 217081 samples with 1000 evaluations per sample.
 Range (min … max):   6.400 ns … 973.200 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     30.100 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   22.028 ns ±  13.612 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

**Float32**
`@benchmark erf(data) setup=(data=6*rand(Float32)-3) samples=1000000`

> BenchmarkTools.Trial: 308269 samples with 1000 evaluations per sample.
 Range (min … max):   4.300 ns … 160.400 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     19.900 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   15.207 ns ±   8.609 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%
 
 # New
**Float64**
`@benchmark erf(data) setup=(data=6*rand(Float64)-3) samples=1000000`

> BenchmarkTools.Trial: 604748 samples with 1000 evaluations per sample.
 Range (min … max):  4.700 ns … 164.400 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     7.400 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   7.302 ns ±   3.017 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

**Float32**
`@benchmark erf(data) setup=(data=6*rand(Float32)-3) samples=1000000`

> BenchmarkTools.Trial: 641285 samples with 1000 evaluations per sample.
 Range (min … max):  4.200 ns … 204.800 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     6.900 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   6.840 ns ±   2.961 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

File: Project.toml [hunk 15]
```diff
@@ -1,6 +1,6 @@
 name = "SpecialFunctions"
 uuid = "276daf66-3868-5448-9aa4-cd146d93841b"
-version = "2.6.1"
+version = "2.6.2"
```
Comment (devmotion) [inline Project.toml]: IMO a Julia implementation of `erf` is a feature, not a bug fix, so the new version should be

```suggestion
version = "2.7.0"
```

**Comment by AhmedYKadah** [comment]
It's been a pleasure 😄
